homeassistant/components/hyperion/light.py

```diff
@@ -157,158 +153,243 @@ def effect(self):
+                    const.KEY_EFFECT: {const.KEY_NAME: effect},
+                    const.KEY_ORIGIN: DEFAULT_ORIGIN,
+                }
+            ):
+                return
+        # == Set a color
+        else:
+            if not await self._client.async_send_set_color(
+                **{
+                    const.KEY_PRIORITY: self._priority,
+                    const.KEY_COLOR: rgb_color,
+                    const.KEY_ORIGIN: DEFAULT_ORIGIN,
+                }
+            ):
+                return
+
+    async def async_turn_off(self, **kwargs):
+        """Disable the LED output component."""
+        _LOGGER.debug("Turning Off: %s", kwargs)
+
+        if not self._client.is_connected:
             return
 
-        cal_color = [int(round(x * float(brightness) / 255)) for x in rgb_color]
-        self.json_request(
-            {"command": "color", "priority": self._priority, "color": cal_color}
-        )
+        if not await self._client.async_send_set_component(
+            **{
+                const.KEY_COMPONENTSTATE: {
+                    const.KEY_COMPONENT: const.KEY_COMPONENTID_LEDDEVICE,
+                    const.KEY_STATE: False,
+                }
+            }
+        ):
+            return
 
-    def turn_off(self, **kwargs):
-        """Disconnect all remotes."""
-        self.json_request({"command": "clearall"})
-        self.json_request(
-            {"command": "color", "priority": self._priority, "color": [0, 0, 0]}
-        )
+    def _set_internal_state(self, brightness=None, rgb_color=None, effect=None):
+        """Set the internal state."""
+        if brightness is not None:
+            self._brightness = brightness
+        if rgb_color is not None:
+            self._rgb_color = rgb_color
+        if effect is not None:
+            self._effect = effect
+            if effect == KEY_EFFECT_SOLID:
+                self._icon = ICON_LIGHTBULB
+            elif effect in const.KEY_COMPONENTID_EXTERNAL_SOURCES:
+                self._icon = ICON_EXTERNAL_SOURCE
+            else:
+                self._icon = ICON_EFFECT
 
-    def update(self):
-        """Get the lights status."""
-        # postpone the immediate state check for changes that take time
-        if self._skip_update:
-            self._skip_update = False
+    def _update_ha_state(self):
+        """Update the internal Home Assistant state."""
+        if not hasattr(self, "hass") or self.hass is None:
             return
-        response = self.json_request({"command": "serverinfo"})
-        if response:
-            # workaround for outdated Hyperion
-            if "activeLedColor" not in response["info"]:
-                self._rgb_color = self._default_color
-                self._rgb_mem = self._default_color
-                self._brightness = 255
-                self._icon = "mdi:lightbulb"
-                self._effect = None
+        self.schedule_update_ha_state()
```

Comment:
I'm assuming the callbacks are called from async context.
```suggestion
        self.async_write_ha_state()
```

Reply:
We can remove this method and call `self.async_write_ha_state` directly instead.